pytorch_lightning/loops/dataloader/evaluation_loop.py

```diff
@@ -272,16 +273,59 @@ def _on_evaluation_epoch_end(self) -> None:
         self.trainer.logger_connector.on_epoch_end()
 
     def _print_results(self, results: List[_OUT_DICT], stage: RunningStage) -> None:
```

Comment:
Perhaps not yet since you are still making changes, but this will need some tests.

Reply:
Yes, agree! What kind of tests did you have in mind and where should I add them? For now, I am still fixing some mypy issues, and thinking about a fix for when multiple metrics are logged under a single name, e.g., 

```python
self.log("performance", {"acc": loss, "recall": loss})
```

because right now the output looks like

```
┏━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃   Metric    ┃    DATALOADER 0     ┃                    DATALOADER 1                     ┃
┡━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ performance │        None         │ {'acc': tensor(-0.5408), 'recall': tensor(-0.5408)} │
│  test_acc   │ -0.8093445301055908 │                 -0.5408046841621399                 │
│  test_loss  │ -0.8093445301055908 │                        None                         │
│   test_x    │ -0.8093445301055908 │                        None                         │
│   test_y    │        None         │                 -0.5408046841621399                 │
│   test_z    │        None         │                 -0.5408046841621399                 │
└─────────────┴─────────────────────┴─────────────────────────────────────────────────────┘
```

Any suggestions here?

Reply:
How does it look if you have 10 dataloaders ?

Reply:
In the latest commit I have modified this bit. Now, in the case of a nested dict, I remove the "performance" as a metric and replace it with all the metrics which it contained (i.e. "acc" and "recall" in the example above). Instead of None I also put a blank right now. Do you agree with these choices?

Reply:
> How does it look if you have 10 dataloaders ?

I have tried this for both the rich and native implementation. For rich, the columns get adjusted to be smaller automatically (until they become unreadable). I will look into whether they can start on a new line. For native, they get really messed up once they do not longer fit the screen, so I will definitely need to fix this.

Reply:
Let's get started on the tests now that the terminal design is basically finalized.

For the tests, you can mock `sys.stdout` and then call `_print_results` with custom inputs and validate that the `stdout` text is the expected.

```python    
from contextlib import redirect_stdout

out = StringIO()
with redirect_stdout(out):
    loop._print_results(...)

expected = ...
assert out.getvalue() == expected
```

Reply:
Thanks for the template, will start on it :)

Reply:
I've added a first test in my latest commit, will continue to work on this.

Reply:
> > How does it look if you have 10 dataloaders ?
> 
> I have tried this for both the rich and native implementation. For rich, the columns get adjusted to be smaller automatically (until they become unreadable). I will look into whether they can start on a new line. For native, they get really messed up once they do not longer fit the screen, so I will definitely need to fix this.

In my latest commit I added functionality for a big native table. For rich, I'm not sure whether it's possible to 'overflow' the table, see: https://rich.readthedocs.io/en/stable/console.html#overflow and https://github.com/Textualize/rich/issues/1357.

Reply:
Not sure why the new tests are failing now, as I believe I have not changed anything which should affect the tests in my latest commit and they passed earlier.

Reply:
We had some CI issues that should be fixed in master. I merged master into your branch, let's see...

Reply:
Still fails, will take a look at it!

Reply:
Was just a small fix of what was expected :) Will take a further look at rich (with 10+ dataloaders) and additional testingv (also mainly for rich). If there are some "exotic" metrics of which you can think, please mention them here? Also should tests be created for the two static methods _get_keys and _find_values? The apply to collection doesnt seem to work (https://github.com/PyTorchLightning/pytorch-lightning/pull/11332#discussion_r780672963).

Reply:
Added logic for large rich tables, I feel like we may shorten the code by not checking whether the table is larger than the terminal. However, then we need to make some decisions on whether we want to add empty cells in the table, and whether to always use the terminal size to draw horizontal lines,... What do you think?

Reply:
> Also should tests be created for the two static methods _get_keys and _find_values? The apply to collection doesnt seem to work (#11332 (comment)).

No need, I'll take a look when reviewing.

> not sure whether this is the best way to suggest alternatives within PRs?

You can comment on your own PR to create separate discussions.

If you want to discuss alternatives, include examples of how it would look like.

Reply:
Okay, thanks!